### PR TITLE
HPCC-18067 Unicode Implementation for Translate

### PR DESCRIPTION
**Pull request by dskaff**
- Adds test cases for TestTranslate in ecllibrary/teststd/uni/
- Adds documentation for stub function in ecllibrary/std/uni.ecl
- Adds the code for the unicode version of translate in unicodelib.cpp

Signed-off-by: David Skaff <david.skaff@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

File: esp/protocols/http/CMakeLists.txt [hunk 1]
```diff
@@ -76,6 +76,7 @@ target_link_libraries(
     ${XALAN_LIBRARIES}
     ${XERCES_LIBRARIES}
     jlib
+	dalibase
```
Comment (richardkchapman) [inline esp/protocols/http/CMakeLists.txt]: This change looks unrelated - it should be in a separate pull request and a separate Jira.

**Comment by dskaff** [comment]
@johnholt please code review.

File: plugins/unicodelib/unicodelib.cpp [hunk 2]
```diff
@@ -719,6 +720,34 @@ unsigned unicodeEditDistanceV4(UnicodeString & left, UnicodeString & right, unsi
...
+    UnicodeString search(sear, searLen);
+    UnicodeString replace(repl, replLen);
+    if (search.countChar32() != replace.countChar32() || source.isEmpty() || search.isEmpty() || replace.isEmpty())
+    {
+        return;
+    }
+    StringCharacterIterator it(source);
+    source.remove();
+    int32_t Beg = it.setToStart();
+    while (Beg != it.endIndex())
```
Comment (johnholt) [inline plugins/unicodelib/unicodelib.cpp]: Without studying the details of CharacterIterator, I cannot tell if this is a problem.  However, it is generally a good practice to use the corresponding pairs:  setToStart() with setToEnd(); or startIndex() with endIndex().  There are sometimes subtle differences.  If you are counting on these differences, you should call that out in a comment.

One minor quibble is that the variable you are setting and updating appears to me to really be a position and not the beginning.

File: ecllibrary/teststd/uni/TestTranslate.ecl [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    //Check action on hebrew lettering with accent marks (called pointing). Bidirectional.
+    //Translation: (not a phrase, 2 different words separated by a space)
+    EXPORT Test33 := ASSERT(Uni.Translate(U'קָמָץ שִׁי״ן',U'קי',U'YX')+U'!' = U'Yָמָץ שִׁX״ן!');
+	//Check action on surrogate pairs.
+    EXPORT Test34 := ASSERT(Uni.Translate(U'x𐐀',U'x𐐀',U'𒀀y')+'!' = U'𒀀y!');
+    EXPORT Test35 := ASSERT(Uni.Translate(U'𐐀',U'𐐀',U'u')+'!' = U'u!');
+    EXPORT Test36 := ASSERT(Uni.Translate(U'x',U'x',U'𐐀')+'!' = U'𐐀!');
+    EXPORT Test37 := ASSERT(Uni.Translate(U'𐐀',U'𐐀',U'xy')+'!' = U'𐐀!');
+	//Check action on combining cedilla to ensure correct Normalization.
+    EXPORT Test38 := ASSERT(Uni.Translate(U'CX',U'X',U'̧')+'!' = U'Ç!');
```
Comment (dskaff) [inline ecllibrary/teststd/uni/TestTranslate.ecl]: Can you give me an example of how you would specify the values in hex in the ECL code?

Also, TRANSLATE doesn't have any normalization processes, I just miss-commented that test case.

File: plugins/unicodelib/unicodelib.cpp (hunk 2 above)
Comment (dskaff) [inline plugins/unicodelib/unicodelib.cpp]: I noticed that they were different as well, but forgot to make them consistent after moving onto issues that were actually a problem for the code's execution. The setTo.. functions are meant for code points, and the ..Index() functions are meant for code units; however, they both set the iterator to the same position in the string, and become irrelevant when the move32 function begins to change the position.

File: ecllibrary/teststd/uni/TestTranslate.ecl (hunk 3 above)
Comment (johnholt) [inline ecllibrary/teststd/uni/TestTranslate.ecl]: Sure, \uyyyy where y is a hex digit.  Example:
UNICODE t0 := u'Test with \ud83c\udf02 (fire) symbol.';

Since TRANSLATE doesn't perform a normalization, shouldn't the locale parameter be dropped?

File: plugins/unicodelib/unicodelib.cpp (hunk 2 above)
Comment (johnholt) [inline plugins/unicodelib/unicodelib.cpp]: Is the value returned by move32 consistent with endIndex?

Comment (dskaff) [inline plugins/unicodelib/unicodelib.cpp]: yes, because both endIndex() and setToEnd() place the iterator on the last possible location for it.

File: ecllibrary/teststd/uni/TestTranslate.ecl (hunk 3 above)
Comment (dskaff) [inline ecllibrary/teststd/uni/TestTranslate.ecl]: Thanks.

And yes, I stopped using it when I moved from a RuleBasedBreakIterator character instance to a StringCharacterIterator, but forgot to remove the actual parameter for it.

File: ecllibrary/std/Uni.ecl [hunk 4]
```diff
@@ -398,4 +398,17 @@ EXPORT unicode GetNthWord(unicode text, unsigned4 n, varstring localename = '')
 EXPORT ExcludeNthWord(unicode text, unsigned4 n, varstring localename = '') :=
     lib_unicodelib.UnicodeLib.UnicodeLocaleExcludeNthWord(text, n, localename);
 
+/**
+ * Returns the source string with the all characters that match characters in the search string replaced
+ * with the character at the corresponding position in the replacement string.
```
Comment (dskaff) [inline ecllibrary/std/Uni.ecl]: The code doesn't try to recognize that. Instead, the search and replacement strings will get passed to the TRANSLATE function where it will first check to make sure that the two strings (now made into Unicode strings) equal each other in character length (character length as in number of code points, not units). I'm not sure how I would check to see if a lone surrogate would bypass this or cause problems.

Comment (johnholt) [inline ecllibrary/std/Uni.ecl]: I would guess that it depends on the Character Iterator that you are using.  You may need to wrap a try/catch on some (or all) of the operations if it throws an exception for an ill-formed UTF-16 string.  Do you already know or should I study the iterator?

Comment (dskaff) [inline ecllibrary/std/Uni.ecl]: The UnicodeString member function that I used to count the code points is countChar32(), and the reference for it says "A code point may occupy either one or two char16_t code units. Counting code points involves reading all code units," suggesting that it might tell when a lone surrogate code unit has been involved.

I could try to implement compareCodePointOrder() instead, which explicitly mentions that the result is not defined if it contained unpaired surrogates.

Comment (johnholt) [inline ecllibrary/std/Uni.ecl]: I don't think that will be necessary.  I need to think through the implications of the description provided by the ICU UnicodeString class for ill formed strings before I can say.

Comment (johnholt) [inline ecllibrary/std/Uni.ecl]: If the string object has an invalid sequence, then isEmpty() will be true.  The explicit test is isBogus().  Since you did not know that and presumably a casual reader will not know that, it would be better to add a comment that your isEmpty() test also tests for invalid sequences.

Comment (dskaff) [inline ecllibrary/std/Uni.ecl]: Okay, I've added the comment.

**Comment by dskaff** [comment]
@johnholt please code review.

File: plugins/unicodelib/unicodelib.cpp [hunk 5]
```diff
@@ -1475,4 +1504,22 @@ UNICODELIB_API void UNICODELIB_CALL ulUnicodeLocaleExcludeNthWord(unsigned & tgt
         tgtLen = 0;
         tgt = 0;
     }
+}
+
+UNICODELIB_API void UNICODELIB_CALL ulUnicodeLocaleTranslate(unsigned & tgtLen, UChar * & tgt, unsigned textLen, UChar const * text, unsigned searLen, UChar const * sear, unsigned replLen, UChar * repl)
+{
+    UErrorCode status = U_ZERO_ERROR;
+    UnicodeString source(text, textLen);
```
Comment (ghalliday) [inline plugins/unicodelib/unicodelib.cpp]: style: source isn't the best name for a string that also returns the processed result - typically you would expect something called source to not be modified.

File: plugins/unicodelib/unicodelib.cpp [hunk 6]
```diff
@@ -1475,4 +1504,22 @@ UNICODELIB_API void UNICODELIB_CALL ulUnicodeLocaleExcludeNthWord(unsigned & tgt
         tgtLen = 0;
         tgt = 0;
     }
+}
+
+UNICODELIB_API void UNICODELIB_CALL ulUnicodeLocaleTranslate(unsigned & tgtLen, UChar * & tgt, unsigned textLen, UChar const * text, unsigned searLen, UChar const * sear, unsigned replLen, UChar * repl)
+{
+    UErrorCode status = U_ZERO_ERROR;
```
Comment (ghalliday) [inline plugins/unicodelib/unicodelib.cpp]: unused variable

File: plugins/unicodelib/unicodelib.cpp [hunk 7]
```diff
@@ -719,6 +720,34 @@ unsigned unicodeEditDistanceV4(UnicodeString & left, UnicodeString & right, unsi
...
+        if (x == -1)
+        {
+            source.append(it.current32());
+        }
+        else
+        {
+            x = replace.moveIndex32(0, x);
+            source.append(replace.char32At(x));
+        }
+        idx = it.move32(1, CharacterIterator::kCurrent);
```
Comment (ghalliday) [inline plugins/unicodelib/unicodelib.cpp]: it.next32() would be clearer, but see note above.

File: plugins/unicodelib/unicodelib.cpp [hunk 8]
```diff
@@ -719,6 +720,34 @@ unsigned unicodeEditDistanceV4(UnicodeString & left, UnicodeString & right, unsi
     return da[mask(leftLen-1)][rightLen-1];
 }
 
+void translate(UnicodeString & source, UChar const * sear, unsigned searLen, UChar const * repl, unsigned replLen)
+{
+    UnicodeString search(sear, searLen);
```
Comment (ghalliday) [inline plugins/unicodelib/unicodelib.cpp]: Efficiency: Could be const for clarity.  Does this constructor clone the input strings?  If yes, then

     UnicodeString search(false, sear, searLen);

may be more efficient.

File: plugins/unicodelib/unicodelib.cpp [hunk 9]
```diff
@@ -719,6 +720,34 @@ unsigned unicodeEditDistanceV4(UnicodeString & left, UnicodeString & right, unsi
...
+    if (search.countChar32() != replace.countChar32() || source.isEmpty() || search.isEmpty() || replace.isEmpty())
+    {
+        return;
+    }
+    StringCharacterIterator it(source);
+    source.remove();
+    int32_t idx = it.setToStart();
+    while (idx != it.endIndex())
+    {
+        int32_t x = search.lastIndexOf(it.current32());
```
Comment (ghalliday) [inline plugins/unicodelib/unicodelib.cpp]: minor: more efficient to have a loop with

    nextCodePoint = it.next32PostInc(), and finish once it equals DONE.

it avoids the call to move32, the check against endIndex.  Also worth avoiding 2 calls to current32.

File: plugins/unicodelib/unicodelib.cpp [hunk 10]
```diff
@@ -719,6 +720,34 @@ unsigned unicodeEditDistanceV4(UnicodeString & left, UnicodeString & right, unsi
...
+            source.append(it.current32());
+        }
+        else
+        {
+            x = replace.moveIndex32(0, x);
+            source.append(replace.char32At(x));
+        }
+        idx = it.move32(1, CharacterIterator::kCurrent);
+    }
+    return;
```
Comment (ghalliday) [inline plugins/unicodelib/unicodelib.cpp]: style: void return at the end of function is redundant.

File: plugins/unicodelib/unicodelib.cpp [hunk 11]
```diff
@@ -719,6 +720,34 @@ unsigned unicodeEditDistanceV4(UnicodeString & left, UnicodeString & right, unsi
     return da[mask(leftLen-1)][rightLen-1];
 }
 
+void translate(UnicodeString & source, UChar const * sear, unsigned searLen, UChar const * repl, unsigned replLen)
```
Comment (ghalliday) [inline plugins/unicodelib/unicodelib.cpp]: style: see comment about "source" as a variable name below.  Something like text or toProcess may be better.

**Review by ghalliday — COMMENTED** [review]
@dskaff this is a good PR.  I *would* take it as it is, but it is worth you looking over the comments.

File: plugins/unicodelib/unicodelib.cpp (hunk 9 above)
Comment (dskaff) [inline plugins/unicodelib/unicodelib.cpp]: When I tried using it.next32PostInc() I ran into issues with the code exiting on the last character without actually running it through because that member returns DONE if the next character is the last character. I'll try to work around it. I think I ran into that problem because I was using the member inside the while loop's condition.

Comment (dskaff) [inline plugins/unicodelib/unicodelib.cpp]: I've edited the code so that it looks like this:

`    while (idx != CharacterIterator::DONE)
    {
        idx = it.next32PostInc();
        int32_t x = search.lastIndexOf(it.current32());
        if (x == -1)
        {
            toProcess.append(it.current32());
        }
        else
        {
            x = replace.moveIndex32(0, x);
            toProcess.append(replace.char32At(x));
        }
    }`

All of the test cases fail with this configuration with the addition of 2 \032 characters to the edited toProcess string.

`ctx->addWuAssertFailure(100000U,"Assert (uvw def \032\032! =  uvw def !) failed [translate(U' xyz abc ', U'xyzabc', U'uvwdef') + U'!' = U' uvw def !']","c:\\users\\skafda01\\hpcc-platform\\ecllibrary\\teststd\\uni\\TestTranslate.ecl",28U,29U,false);
				ctx->addWuAssertFailure(100000U,"Assert (uvw abc \032\032! =  uvw abc !) failed [translate(U' xyz abc ', U'zyx', U'wvu') + U'!' = U' uvw abc !']","c:\\users\\skafda01\\hpcc-platform\\ecllibrary\\teststd\\uni\\TestTranslate.ecl",29U,29U,false);`

I'm not sure why this is happening.

**Comment by AttilaVamos** [comment]
Schedule2.ecls error is unrelated to this PR. It is rarely appearing then disappears. I rescheduled the Smoketest.

**Comment by dskaff** [comment]
@ghalliday please code review

File: plugins/unicodelib/unicodelib.cpp (hunk 9 above)
Comment (ghalliday) [inline plugins/unicodelib/unicodelib.cpp]: This is what the code would look like if you are using next32PostInc():

    it.setToStart();
    for(;;)
    {
        int32_t next = it.next32PostInc();
        if (next == CharacterIterator::DONE)
            break;
        int32_t x = search.lastIndexOf(next);
        if (x == -1)
        {
            toProcess.append(next);
        }
        else
        {
            x = replace.moveIndex32(0, x);
            toProcess.append(replace.char32At(x));
        }
    }

File: plugins/unicodelib/unicodelib.cpp [hunk 12]
```diff
@@ -719,6 +720,33 @@ unsigned unicodeEditDistanceV4(UnicodeString & left, UnicodeString & right, unsi
     return da[mask(leftLen-1)][rightLen-1];
 }
 
+void translate(UnicodeString & toProcess, UChar const * sear, unsigned searLen, UChar const * repl, unsigned replLen)
+{
+    UnicodeString search(false, sear, searLen);
+    UnicodeString replace(repl, replLen);
```
Comment (ghalliday) [inline plugins/unicodelib/unicodelib.cpp]: You could have also change the constructor for this string, and made them both const to emphasises they never changed.

**Comment by ghalliday** [comment]
@dskaff I added a couple of comments - but will merge anyway.

**Comment by ghalliday** [comment]
This is what the code would look like if you are using next32PostInc():

    it.setToStart();
    for(;;)
    {
        int32_t next = it.next32PostInc();
        if (next == CharacterIterator::DONE)
            break;
        int32_t x = search.lastIndexOf(next);
        if (x == -1)
        {
            toProcess.append(next);
        }
        else
        {
            x = replace.moveIndex32(0, x);
            toProcess.append(replace.char32At(x));
        }
    }
